draft-ietf-tls-tls13.md

```diff
@@ -3450,60 +3409,74 @@ the sources from the table above.
-                             handshake_hash, L)
-
-  4. mES = HKDF-Expand-Label(xES, "expanded ephemeral secret",
-                             handshake_hash, L)
-
-  5. master_secret = HKDF-Extract(mSS, mES)
-
-
-  6. traffic_secret_0 = HKDF-Expand-Label(master_secret,
-                                          "traffic secret",
-                                          handshake_hash, L)
-
-  Where handshake_hash includes all messages up through the
-  server CertificateVerify message, but not including any
-  0-RTT handshake messages (the server's Finished is not
-  included because the master_secret is needed to compute
-  the finished key). [[OPEN ISSUE: Should we be including
-  0-RTT handshake messages here and below?.
-  https://github.com/tlswg/tls13-spec/issues/351
-  ]] At this point,
-  SS, ES, xSS, xES, mSS, and mES SHOULD be securely deleted,
-  along with any ephemeral (EC)DH secrets.
-
-  7. resumption_secret = HKDF-Expand-Label(master_secret,
-                                           "resumption master secret"
-                                           handshake_hash, L)
-
-  8. exporter_secret = HKDF-Expand-Label(master_secret,
-                                         "exporter master secret",
-                                         handshake_hash, L)
-
-
-  Where handshake_hash contains the entire handshake up to
-  and including the client's Finished, but not including
-  any 0-RTT handshake messages or post-handshake messages.
-  AT this point master_secret SHOULD be securely deleted.
-~~~
+  Add-Secret(Secret, InputSecret) =
+       HKDF-Extract(HKDF-Expand-Label(Secret, "add secret", "", L),
+                                      InputSecret))
+
+  where L denotes the length of the underlying hash function for
+  HKDF.
+
+  Derive-Secret(Secret, Label, Messages) =
+       HKDF-Extract(HKDF-Expand(Secret, Label, Hash(Messages), L))
+~~~~       
+
+Given a set of n InputSecrets, the final "master secret" is computed
+by iteratively invoking Add-Secret with InputSecret_1, InputSecret_2,
+etc.  The initial secret is simply a string of 0s as long as the size
+of the Hash that is the basis for the HKDF. Concretely, for the
+present version of TLS 1.3, secrets are added in the following order:
+
+- PSK
+- (EC)DHE shared secret
+
+This produces a full key derivation schedule as below.
+
+~~~~ 
+                 0
+                 |
+   PSK ->   Add-Secret()
+                 |
+                 v
+           Early Secret  --> Derive-Secret("early traffic secret",
+                 |                         ClientHello) -> early_traffic_secret
+                 |                         
+                 v
+(EC)DHE ->  Add-Secret()
+                 |
+                 v
+           Master Secret --> Derive-Secret("handshake traffic secret",
+                 |                         ClientHello + ServerHello) ->
+                 |                         handshake_traffic_secret
```

Comment:
I'm not sure that this formatting works that well, the handshake_traffic_secret looks like it is an argument to Derive-Secret with this indent.